language/Type-Classes.md

```diff
@@ -25,6 +25,28 @@ instance showArray :: (Show a) => Show (Array a) where
 example = show [true, false]
 ```
 
+## Instance selection
+
+Instance selection works by looking at the "head" types of each type class parameter. Functional dependencies also plays a role, but we'll not consider that for now to simplify the explanation. The "head" of a type is the top-most type constructor, for example the head of `Array Int` is `Array` and the head of `Record r` is `Record`.
+
+``` purescript
+class C t where
+  f :: t -> String
+-- Instances are chosen 
+instance cRec :: C (Record r) where
+  f _ = "this is a record"
+instance cArr :: C (Array a) where
+  f _ = "this is an array"
+
+main = do
+  -- The `f` function resolves to the instance for the argument type's head.
+  log (f {}) -- Resolves to cRec's f.
+  log (f []) -- Resolves to cArr's f.
+-- Prints the following when executed.
+this is a record
+this is an array
+```
```

Comment:
I don't think this is right. For instance:

```purescript
module Main where

import Prelude
import Data.Foldable (fold)
import TryPureScript

main = render $ fold $
          [ p $ text $ describe [1, 2, 3]
          , p $ text $ describe [1.0, 2.0, 3.0]
          ]

class Describe a where
  describe :: a -> String
  
instance describe1 :: Describe (Array Int) where
  describe _ = "I am an Array Int"
  
instance describe2 :: Describe (Array Number) where
  describe _ = "I am an Array Number"
```

The compiler has had to look beyond the head of each instance to select it here.